src/contract.rs

```diff
@@ -254,13 +311,36 @@ pub fn execute_mint_path(
     path: String,
 ) -> Result<Response, ContractError> {
     let config = CONFIG.load(deps.storage)?;
-    let payment_details = PAYMENT_DETAILS.may_load(deps.storage)?;
+    let mut payment_details = PAYMENT_DETAILS.may_load(deps.storage)?;
     if config.token_id.is_none() {
         // No token to mint off of
         return Err(ContractError::NoRootToken {});
     }
     let token_id = config.token_id.unwrap();
 
+    let is_in_claim_window_ = is_in_claim_window(
+        config.reserve_root_for_n_blocks,
+        config.initial_height,
+        env.block.height,
+    );
+
+    if is_in_claim_window_ {
+        if !config.reserve_root_names {
+            return Err(ContractError::ReserveRootNameDisabled {});
+        }
+
+        let path_as_base_owner =
+            get_dens_owner(&deps.querier, path.clone(), config.whoami_address.clone());
+
+        if let Some(path_as_base_owner) = path_as_base_owner {
+            if info.sender.to_string() == path_as_base_owner {
+                payment_details = None
+            } else {
+                return Err(ContractError::RootInClaimWindowToken {});
+            }
+        }
+    }
+
     if let Some(payment_details) = payment_details {
```

Comment:
Then this line is something like `Some(payment_details) && !override_payment`

Reply:
I bumped into this https://github.com/rust-lang/rust/issues/53667 so I write it a bit different